### PR TITLE
Update generators to optionally generate files with lower case path.

### PR DESCRIPTION
**Pull request by Astro03**
**Issue**:
Was working on two differnet file systms and noticed that after a rync,
I was getting a "DuplicateClassException" on my Linux system - there
were dupiclate files in similar paths.

Typically a Mac/Windows system will have a case insensitive file system,
whereas a Linux system will have a case sensitive file system. For a
case insensitive file system, "\~/com/astro" and "\~/com/ASTRO" point to
the same folder. For a case sensitive file system, "\~/com/astro" and
"\~/com/ASTRO" will be different folders.

 **Example**:
   File1: namespace = com.astro.file1
   File2: namespace = com.aSTRo.file2

   The following files would be generated with the path specified.
   1) Case insensitive (if file1 is generated first):
       com/astro/file1
                /file2
   2) Case insensitive (if file2 is generated first):
       com/aSTRo/file1
                /file2
   3) Case sensitive:
       com/astro/file1
           aSTRo/file2

**Resolution**:
Add abililty for generators to generate paths with lower
case path. Updated the TempalteSpecGenerator,
PegasusDataTemplateGenerator, RestRequestBuilderGenerator, and
JavaRequestBuilderGenerator.

The Gradle Tasks of GenerateDataTemplateTask and
GenerateRestClientTask have been modified to have an optional input property
to generate teh lower case path.

**Testing**:
Created unit tests to validate path's on both linux and mac systems
where the file system were case sensitive and case insensitive
respectively.

### CODE REVIEW

**Comment by Astro03** [comment]
@karthikbalasub @junchuanwang  
Can you both please look at this and see if I missed anything. Thank you.

File: generator/src/main/java/com/linkedin/pegasus/generator/JavaCodeUtil.java [hunk 1]
```diff
@@ -146,7 +157,17 @@ else if (!checker.isPersistent(definedClass))
         }
         else if (definedClass.outer() == null)
         {
-          final File file = new File(targetDirectory, definedClass.fullName().replace('.', File.separatorChar) + ".java");
+          String path;
+          if (generateLowercasePath) {
+            // Create path this way since fullName() has a recursive call.
+            String fullName = definedClass.fullName();
+            String name = definedClass.name();
+            String packageName = fullName.substring(0, fullName.length() - name.length());
```
Comment (karthikbalasub) [inline generator/src/main/java/com/linkedin/pegasus/generator/JavaCodeUtil.java:165]: Use definedClass.getPackage().name() instead?

File: generator/src/main/java/com/linkedin/pegasus/generator/PegasusDataTemplateGenerator.java [hunk 2]
```diff
@@ -49,6 +44,7 @@
    * The system property that specifies whether to generate classes for externally resolved schemas
    */
   public static final String GENERATOR_GENERATE_IMPORTED = "generator.generate.imported";
+  public static final String GENERATOR_GENERATE_PATH_LOWERCASE = "generator.generate.path.lowercase";
```
Comment (karthikbalasub) [inline generator/src/main/java/com/linkedin/pegasus/generator/PegasusDataTemplateGenerator.java:47]: rename for consistency:
PATH_LOWERCASE -> LOWERCASE_PATH

File: generator/src/main/java/com/linkedin/pegasus/generator/JavaCodeUtil.java [hunk 3]
```diff
@@ -17,26 +17,22 @@
...
 import java.net.URI;
 import java.net.URL;
 import java.net.URLClassLoader;
+import java.nio.file.Paths;
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
 import java.util.StringTokenizer;
 
-import com.sun.codemodel.JAnnotationUse;
```
Comment (karthikbalasub) [inline generator/src/main/java/com/linkedin/pegasus/generator/JavaCodeUtil.java:32]: Revert these changes, pegasus codestyle doesn't allow wildcard imports (except for static methods)

File: generator/src/main/java/com/linkedin/pegasus/generator/TemplateSpecGenerator.java [hunk 4]
```diff
@@ -635,7 +624,8 @@ private ClassTemplateSpec generateUnion(UnionDataSchema schema, TyperefDataSchem
     pushCurrentLocation(_schemaResolver.nameToDataSchemaLocations().get(typerefDataSchema.getFullName()));
 
     final UnionTemplateSpec unionClass = new UnionTemplateSpec(schema);
-    unionClass.setNamespace(typerefDataSchema.getNamespace());
+    String namespace = _lowercaseNamespace ? typerefDataSchema.getNamespace().toLowerCase() : typerefDataSchema.getNamespace();
+    unionClass.setNamespace(namespace);
```
Comment (karthikbalasub) [inline generator/src/main/java/com/linkedin/pegasus/generator/TemplateSpecGenerator.java:628]: This is changing the namespace on the generate template classes and it is a breaking change. This config should only change the generated directories

File: restli-common/src/main/java/com/linkedin/restli/restspec/RestSpecCodec.java [hunk 5]
```diff
@@ -81,8 +80,27 @@ public RestSpecCodec()
...
+   * @throws IOException
+   */
+  public ResourceSchema readResourceSchema(InputStream inputStream, boolean generateLowercasePath)
+          throws IOException{
     final DataMap data = _dataCodec.readMap(inputStream);
     fixupLegacyRestspec(data);
+    if (generateLowercasePath && data.containsKey("namespace")) {
+      String namespace = data.getString("namespace");
+      if (namespace != null) {
+        data.put("namespace", namespace.toLowerCase());
```
Comment (karthikbalasub) [inline restli-common/src/main/java/com/linkedin/restli/restspec/RestSpecCodec.java:101]: Same here, avoid changing the namespace of the schema

File: restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/RestRequestBuilderGenerator.java [hunk 6]
```diff
@@ -54,6 +47,7 @@
 {
   static final String GENERATOR_REST_GENERATE_DATATEMPLATES = "generator.rest.generate.datatemplates";
   static final String GENERATOR_REST_GENERATE_VERSION = "generator.rest.generate.version";
+  public static final String GENERATOR_REST_GENERATE_PATH_LOWERCASE = "generator.rest.generate.path.lowercase";
```
Comment (karthikbalasub) [inline restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/RestRequestBuilderGenerator.java:50]: LOWERCASE_PATH

**Review by karthikbalasub — CHANGES_REQUESTED** [review]

File: generator/src/main/java/com/linkedin/pegasus/generator/JavaCodeUtil.java (hunk 1 above)
Comment (Astro03) [reply generator/src/main/java/com/linkedin/pegasus/generator/JavaCodeUtil.java:165]: Did you read the comment? I'm not usually in this codebase but I saw that there was a recursive call in the way fullName worked and thus did it this way. Obviously creating it with the package and doing tolower there would be faster but i wasn't sure if it would produce the same result. 

thoughts?

Comment (Astro03) [reply generator/src/main/java/com/linkedin/pegasus/generator/JavaCodeUtil.java:165]: @karthikbalasub ^

File: generator/src/main/java/com/linkedin/pegasus/generator/PegasusDataTemplateGenerator.java (hunk 2 above)
Comment (Astro03) [reply generator/src/main/java/com/linkedin/pegasus/generator/PegasusDataTemplateGenerator.java:47]: My original thought was that having grouping of "path" with leaf node "lowercase" made more sense than having a "lowercase" grouping with "path" as a leaf node. 

I will change if this didn't  logic didn't make sense to you.

File: generator/src/main/java/com/linkedin/pegasus/generator/JavaCodeUtil.java (hunk 3 above)
Comment (Astro03) [reply generator/src/main/java/com/linkedin/pegasus/generator/JavaCodeUtil.java:32]: Oh okay. I should do that to all the files. These were done by IntelliJ.  I'll do it for all files. Do we have a linter for this that would catch this issue?

Comment (Astro03) [reply generator/src/main/java/com/linkedin/pegasus/generator/JavaCodeUtil.java]: reverted. 

I do have concern about needing to change the codestyle since IntelliJ natively did some of these changes without my intervention.

File: generator/src/main/java/com/linkedin/pegasus/generator/JavaCodeUtil.java (hunk 1 above)
Comment (karthikbalasub) [reply generator/src/main/java/com/linkedin/pegasus/generator/JavaCodeUtil.java:169]: Let us keep it this way to be safe.

File: restli-common/src/main/java/com/linkedin/restli/restspec/RestSpecCodec.java (hunk 5 above)
Comment (Astro03) [reply restli-common/src/main/java/com/linkedin/restli/restspec/RestSpecCodec.java:102]: I would love to get on a call and discuss this with you. 

If i remove this, then the the paths will not change for the builders.

**Comment by Astro03** [comment]
Closing since we will go with a different approach. Will change the the file writer class directly -> smarter approach.

**Comment by Astro03** [comment]
Created these PRs with different approaches. 

https://github.com/linkedin/rest.li/pull/473 - all generated paths are lower case
https://github.com/linkedin/rest.li/pull/474 - default generated paths are lower case; use optional flag to make it as it was.
